File: asm/src/ist0_set.py
from instruction import InstructionPrototype
from parameter import Parameter


ist0_set = [
    InstructionPrototype(0x00, 'nop',
                         'no operation',
                         [],
                         0, 0),
    InstructionPrototype(0x01, 'i32',
                         'put one 32 bit integer on stack',
                         [Parameter('val', 'integer value')],
                         0, 1),
    InstructionPrototype(0x02, 'add',
                         'add two 32 bit integers',
                         [],
                         2, 1),
    InstructionPrototype(0x03, 'dec',
                         'decrement top of the stack',
                         [],
                         0, 0),
    InstructionPrototype(0x10, 'eq?',
                         'test if two 32 bit integers, producing 0/1 value',
                         [],
                         2, 1),
    InstructionPrototype(0x11, 'ez?',
                         'test if top of stack is zero, producing 0/1 value',
                         [],
                         0, 1),
    InstructionPrototype(0x20, 'jez',
                         'jump to address if 0 is popped from the stack',
                         [Parameter('addr', 'jump address')],
                         1, 0),
    InstructionPrototype(0x21, 'jnz',
                         'jump to address if 1 is popped from the stack',
                         [Parameter('addr', 'jump address')],
                         1, 0)
]

File: vm/src/ist0_set.c
#include "ist0_set.h"
#include "instr.h"

INSTR_FN(nop) {}

INSTR_FN(i32)
{
	PRODUCE(*pv);
	NEXT_LOC();
}

INSTR_FN(add)
{
	int32_t a, b;
	CONSUME(&a);
	CONSUME(&b);
	PRODUCE(a + b);
	NEXT_LOC();
}

INSTR_FN(dec)
{
	int32_t a;
	CONSUME(&a);
	PRODUCE(a - 1);
	NEXT_LOC();
}

INSTR_FN(eq_)
{
	int32_t a, b;
	CONSUME(&a);
	CONSUME(&b);
	PRODUCE(a == b);
	NEXT_LOC();
}

INSTR_FN(ez_)
{
	int32_t a;
	CONSUME(&a);
	PRODUCE(a);
	PRODUCE(a == 0);
	NEXT_LOC();
}

INSTR_FN(jez)
{
	int32_t a;
	CONSUME(&a);
	if (a == 0) SET_LOC(*pv);
	else NEXT_LOC();
}

INSTR_FN(jnz)
{
	int32_t a;
	CONSUME(&a);
	if (a != 0) SET_LOC(*pv);
	else NEXT_LOC();
}

struct instr_def ist0_set[MAX_SET] =
{
	INSTR_ENTRY(0x00, nop, 0),
	INSTR_ENTRY(0x01, i32, 1),
	INSTR_ENTRY(0x02, add, 0),
	INSTR_ENTRY(0x03, dec, 0),
	INSTR_ENTRY(0x10, eq_, 0),
	INSTR_ENTRY(0x11, ez_, 0),
	INSTR_ENTRY(0x20, jez, 1),
	INSTR_ENTRY(0x21, jnz, 1),
};

